File: src/graphics/camera_debug.cpp
#include "city/city.h"

#include "widget/debug_console.h"
#include "imgui/imgui.h"
#include "imgui/imgui_internal.h"
#include "view/lookup.h"
#include "game/game.h"

ANK_REGISTER_PROPS_ITERATOR(config_show_camera_properties);
void config_show_camera_properties(bool header) {
    if (header) {
        return;
    }

    bool common_open = ImGui::TreeNodeEx("Camera", ImGuiTreeNodeFlags_None, "Camera");
    if (common_open) {
        ImGui::BeginTable("Camera", 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

        painter ctx = game.painter();

        tile2i camera_tile = city_view_get_camera_mappoint();
        vec2i camera_pixels = camera_get_pixel_offset_internal(ctx);

        auto mm_view = g_city_view.get_scrollable_pixel_limits();

        vec2i real_max = city_view_get_camera_max_tile();

        vec2i max_pixel_offset = city_view_get_camera_max_pixel_offset();

        game_debug_show_property("view min", mm_view.min);
        game_debug_show_property("view max", mm_view.max);
        game_debug_show_property("camera", g_city_view.camera.position);

        game_debug_show_property("scroll min", vec2i{ SCROLL_MIN_SCREENTILE_X, SCROLL_MIN_SCREENTILE_Y }, true);
        game_debug_show_property("tile", camera_tile);
        game_debug_show_property("rela_max", real_max);

        game_debug_show_property("pixel min", vec2i{ 0, 0 }, true);
        game_debug_show_property("pixel", camera_pixels);
        game_debug_show_property("pixek max", max_pixel_offset);

        game_debug_show_property("v.tiles", vec2i{ g_city_view.viewport.size_pixels.x / 60, g_city_view.viewport.size_pixels.y / 30 }, true);
        game_debug_show_property("v.pixels", g_city_view.viewport.size_pixels);

        game_debug_show_property("zoom", g_zoom.get_percentage(), true);
        game_debug_show_property("scale", g_zoom.get_scale(), true);

        game_debug_show_property("target", g_zoom.ftarget(), true);
        game_debug_show_property("delta", g_zoom.fdelta(), true);

        vec2i pixel = { mouse_get()->x, mouse_get()->y };
        game_debug_show_property("mouse", pixel, true);

        vec2i viewp = pixel_to_viewport(pixel);
        game_debug_show_property("viewp", viewp, true);

        vec2i coord = pixel_to_camera_coord(pixel, false);
        game_debug_show_property("coord", coord, true);

        screen_tile screen = pixel_to_screentile(pixel);
        game_debug_show_property("tile", screen);

        vec2i offset = { coord.x % TILE_WIDTH_PIXELS, coord.y % TILE_HEIGHT_PIXELS };
        game_debug_show_property("offset", offset, true);
        
        tile2i point = screen_to_tile(screen);
        game_debug_show_property("point", point);
        game_debug_show_property("grid_offset", point.grid_offset(), true);

        game_debug_show_property("terrain type", get_terrain_type("type: ", point), true);
        pixel = lookup_tile_to_pixel(point);
        game_debug_show_property("pixel", pixel, true);
       
        ImGui::EndTable();

        ImGui::TreePop();
    }
}

File: src/scripts/camera_debug.js
log_info("akhenaten: camera_debug")

var TILE_WIDTH_PIXELS = 60
var TILE_HEIGHT_PIXELS = 30

[es=event_draw_debug_properties]
function camera_debug_draw_properties(ev) {
    if (!imgui.tree_node_ex("Camera")) {
        return
    }

    imgui.begin_table("Camera", 2, imgui.table_flags_debug_props())

    imgui.property_input("view min", __city_view_scroll_min())
    imgui.property_input("view max", __city_view_scroll_max())
    imgui.property_input("camera", __city_view, "camera_position")

    imgui.property_input("scroll min", __city_view, "scroll_min_screentile")
    imgui.property_input("tile", __city_view, "camera_mappoint")
    imgui.property_input("rela_max", __city_view, "camera_max_tile")

    imgui.property_input("pixel min", Vec2i(0, 0))
    imgui.property_input("pixel", __city_view, "camera_pixel_offset_internal")
    imgui.property_input("pixek max", __city_view, "camera_max_pixel_offset")

    imgui.property_input("v.tiles", __city_view, "viewport_tiles")
    imgui.property_input("v.pixels", __city_view, "size_pixels")

    imgui.property_input("zoom", __zoom_percentage())
    imgui.property_input("scale", __zoom_scale())
    imgui.property_input("target", __zoom_target())
    imgui.property_input("delta", __zoom_delta())

    var pixel = Vec2i(__mouse.x, __mouse.y)
    imgui.property_input("mouse", pixel)
    imgui.property_input("viewp", __city_view, "mouse_viewport")
    imgui.property_input("coord", __city_view, "mouse_camera_coord")

    var screen = __pixel_to_screentile(pixel)
    imgui.property_input("tile", screen)

    imgui.property_input("offset", __city_view, "mouse_camera_offset")

    var point = __screen_to_tile(screen)
    imgui.property_input("point", point)
    imgui.property_input("grid_offset", __city_view, "mouse_grid_offset")

    imgui.property_input("terrain type", __city_view, "mouse_terrain_type")
    imgui.property_input("pixel", __city_view, "mouse_tile_pixel")

    imgui.end_table()
    imgui.tree_pop()
}
